From b6d7d6fa782e2e08a47e45b7aba1f457ae7ee008 Mon Sep 17 00:00:00 2001
From: Shivaprasad Hongal <shongal@codeaurora.org>
Date: Wed, 10 Oct 2018 16:05:33 -0700
Subject: [PATCH 3/4] vold: add support for clear key

Add support for clearing key in trustzone and ICE
when a user is deleted.

CRs-Fixed: 2334473
Change-Id: I5fd75aaa3b35145b744bed384dc3c842185ff267
---
 Ext4Crypt.cpp | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/Ext4Crypt.cpp b/Ext4Crypt.cpp
index 6ab1d0a..6d6adee 100644
--- a/Ext4Crypt.cpp
+++ b/Ext4Crypt.cpp
@@ -484,7 +484,6 @@ static void drop_caches() {
 }
 
 static bool evict_ce_key(userid_t user_id) {
-    s_ce_keys.erase(user_id);
     bool success = true;
     std::string raw_ref;
     // If we haven't loaded the CE key, no need to evict it.
@@ -492,6 +491,23 @@ static bool evict_ce_key(userid_t user_id) {
         success &= android::vold::evictKey(raw_ref);
         drop_caches();
     }
+
+    if(is_wrapped_key_supported()) {
+        KeyBuffer key;
+        key = s_ce_keys[user_id];
+
+        std::string keystr(key.data(), key.size());
+        Keymaster keymaster;
+
+        if (!keymaster) {
+            s_ce_keys.erase(user_id);
+            s_ce_key_raw_refs.erase(user_id);
+            return false;
+        }
+        keymaster.deleteKey(keystr);
+    }
+
+    s_ce_keys.erase(user_id);
     s_ce_key_raw_refs.erase(user_id);
     return success;
 }
-- 
2.17.1

